trin-metrics/src/timer.rs

```diff
@@ -0,0 +1,266 @@
+        assert!(proto_histogram.get_sample_sum() >= 0.1);
+    }
+
+    #[test]
+    fn test_stop_and_record_in_thread() {
+        let opts = HistogramOpts::new("test_stop_and_record_in_thread", "testing")
+            .const_label("defense", "1");
+        let histogram = Histogram::with_opts(opts).unwrap();
+
+        let timer = DiscardOnDropHistogramTimer::new(histogram.clone());
+        let handler = thread::spawn(move || {
+            let timer = timer;
+            thread::sleep(Duration::from_millis(100));
+            let time = timer.stop_and_record();
+            assert!(time >= 0.1);
+        });
+        assert!(handler.join().is_ok());
+
+        let mut metric_families = histogram.collect();
+        assert_eq!(metric_families.len(), 1);
+
+        let metric_family = metric_families.pop().unwrap();
+        let metric = metric_family.get_metric().first().unwrap();
+        assert_eq!(metric.get_label().len(), 1);
+        let proto_histogram = metric.get_histogram();
+        assert_eq!(proto_histogram.get_sample_count(), 1);
+        assert!(proto_histogram.get_sample_sum() >= 0.1);
+    }
+
+    #[test]
+    fn test_stop_and_record() {
+        let opts =
+            HistogramOpts::new("test_stop_and_record", "testing").const_label("defense", "1");
+        let histogram = Histogram::with_opts(opts).unwrap();
+
+        let timer = DiscardOnDropHistogramTimer::new(histogram.clone());
+        thread::sleep(Duration::from_millis(100));
+        let time = timer.stop_and_record();
+        assert!(time >= 0.1);
+
+        let mut metric_families = histogram.collect();
+        assert_eq!(metric_families.len(), 1);
+
+        let metric_family = metric_families.pop().unwrap();
+        let metric = metric_family.get_metric().first().unwrap();
+        assert_eq!(metric.get_label().len(), 1);
+        let proto_histogram = metric.get_histogram();
+        assert_eq!(proto_histogram.get_sample_count(), 1);
+        assert!(proto_histogram.get_sample_sum() >= 0.1);
+    }
+
+    #[test]
+    fn test_stop_and_discard() {
+        let opts =
+            HistogramOpts::new("test_stop_and_discard", "testing").const_label("defense", "1");
+        let histogram = Histogram::with_opts(opts).unwrap();
+
+        let timer = DiscardOnDropHistogramTimer::new(histogram.clone());
+        thread::sleep(Duration::from_millis(100));
+        let time = timer.stop_and_discard();
+        assert!(time >= 0.1);
+
+        let mut metric_families = histogram.collect();
+        assert_eq!(metric_families.len(), 1);
+
+        let metric_family = metric_families.pop().unwrap();
+        let metric = metric_family.get_metric().first().unwrap();
+        assert_eq!(metric.get_label().len(), 1);
+        let proto_histogram = metric.get_histogram();
+        assert_eq!(proto_histogram.get_sample_count(), 0);
+        assert!(proto_histogram.get_sample_sum() >= 0.0);
+    }
+
+    #[test]
+    fn test_discard_through_explict_drop() {
```

Comment:
`explicit` & below in the histogram name. and `implicit` for the same cases in the tests below